File: backend/app/routes/post.py
import uuid
from datetime import datetime
from fastapi import APIRouter, Depends, UploadFile, File, Form, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.user import User
from app.models.instagram_account import InstagramAccount
from app.models.post_history import PostHistory
from app.schemas.api import ApiResponse
from app.schemas.post import PostPublish, PostResponse
from app.services.image_service import image_service
from app.services.instagram_service import instagram_service
from app.routes.deps import get_current_user
from app.lib.errors import NotFoundError, ForbiddenError, InstagramError

router = APIRouter(prefix="/posts", tags=["posts"])

class ImageUploadResponse(BaseModel):
    imageId: str

@router.post("/upload", response_model=ApiResponse[ImageUploadResponse], status_code=201)
async def upload_post_image(
    image: UploadFile = File(...),
    current_user: User = Depends(get_current_user)
):
    """画像を一時保存する"""
    image_id = image_service.save_image(image, current_user.id)
    return ApiResponse(data=ImageUploadResponse(imageId=image_id))

@router.post("/publish", response_model=ApiResponse[PostResponse])
def publish_post(
    data: PostPublish,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Instagramに写真を投稿する"""
    # アカウントの存在確認と権限チェック
    account = db.query(InstagramAccount).filter(InstagramAccount.id == data.instagramAccountId).first()
    if not account:
        raise NotFoundError("Instagramアカウントが見つかりません")
    
    if account.user_id != current_user.id:
        raise ForbiddenError("このInstagramアカウントを使用する権限がありません")

    # 投稿履歴のレコード作成
    post_id = str(uuid.uuid4())
    db_post = PostHistory(
        id=post_id,
        user_id=current_user.id,
        instagram_account_id=account.id,
        caption=data.caption,
        status="uploading"
    )
    db.add(db_post)
    db.commit()
    db.refresh(db_post)

    try:
        # 画像パスの取得
        image_path = image_service.get_image_path(current_user.id, data.imageId)
        
        # 投稿開始
        db_post.status = "publishing"
        db.commit()

        # instagrapiで写真アップロード
        media_id = instagram_service.upload_photo(
            session_data_encrypted=account.session_data,
            image_path=image_path,
            caption=data.caption
        )

        # 投稿成功処理
        db_post.status = "success"
        db_post.instagram_post_id = media_id
        db_post.posted_at = datetime.utcnow()
        db.commit()
        db.refresh(db_post)

        # 一時画像の削除
        image_service.delete_image(current_user.id, data.imageId)

        return ApiResponse(data=PostResponse.model_validate(db_post))

    except Exception as e:
        # 投稿失敗処理
        db_post.status = "failed"
        db.commit()
        if isinstance(e, InstagramError):
            raise e
        raise InstagramError(f"Instagramへの投稿に失敗しました: {str(e)}")

@router.get("/status/{post_id}", response_model=ApiResponse[PostResponse])
def get_post_status(
    post_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """投稿のステータスを取得する"""
    post = db.query(PostHistory).filter(PostHistory.id == post_id).first()
    if not post:
        raise NotFoundError("投稿が見つかりません")
        
    if post.user_id != current_user.id:
        raise ForbiddenError("この投稿情報にアクセスする権限がありません")
        
    return ApiResponse(data=PostResponse.model_validate(post))

File: worker/src/routes/post.ts
import { Hono } from 'hono';
import type { Env } from '../types/env';
import { authMiddleware } from '../middleware/auth';
import { uploadImage, getImage, deleteImage, buildImageUrl } from '../services/imageService';
import { createMediaContainer, publishMedia } from '../services/instagramApi';
import { NotFoundError, ForbiddenError, ValidationError } from '../lib/errors';

type PostEnv = { Bindings: Env; Variables: { userId: string } };

const post = new Hono<PostEnv>();

post.use('*', authMiddleware);

/** POST /api/posts/upload - 画像アップロード（R2へ保存） */
post.post('/upload', async (c) => {
  const userId = c.get('userId');
  const formData = await c.req.formData();
  const file = formData.get('image');

  if (!(file instanceof File)) throw new ValidationError('画像ファイルが必要です');

  const arrayBuffer = await file.arrayBuffer();
  const imageId = await uploadImage(c.env.BUCKET, userId, arrayBuffer, file.type);

  return c.json({ success: true, data: { imageId } }, 201);
});

/** POST /api/posts/publish - Instagram投稿実行 */
post.post('/publish', async (c) => {
  const userId = c.get('userId');
  const body = await c.req.json<{
    imageId: string;
    instagramAccountId: string;
    caption: string;
  }>();

  const account = await c.env.DB
    .prepare('SELECT id, user_id, instagram_user_id, access_token_encrypted FROM instagram_accounts WHERE id = ?')
    .bind(body.instagramAccountId)
    .first<{
      id: string;
      user_id: string;
      instagram_user_id: string;
      access_token_encrypted: string;
    }>();

  if (!account) throw new NotFoundError('Instagramアカウントが見つかりません');
  if (account.user_id !== userId) throw new ForbiddenError();

  const postId = crypto.randomUUID();
  await c.env.DB
    .prepare(
      'INSERT INTO post_history (id, user_id, instagram_account_id, caption, status) VALUES (?, ?, ?, ?, ?)'
    )
    .bind(postId, userId, account.id, body.caption, 'uploading')
    .run();

  const workerUrl = new URL(c.req.url).origin;
  const imageUrl = buildImageUrl(workerUrl, body.imageId);

  try {
    await c.env.DB
      .prepare('UPDATE post_history SET status = ? WHERE id = ?')
      .bind('publishing', postId)
      .run();

    const containerId = await createMediaContainer(
      account.instagram_user_id,
      imageUrl,
      body.caption,
      account.access_token_encrypted
    );

    const igPostId = await publishMedia(
      account.instagram_user_id,
      containerId,
      account.access_token_encrypted
    );

    await c.env.DB
      .prepare('UPDATE post_history SET status = ?, instagram_post_id = ?, posted_at = datetime(\'now\') WHERE id = ?')
      .bind('success', igPostId, postId)
      .run();

    await deleteImage(c.env.BUCKET, userId, body.imageId);

    return c.json({
      success: true,
      data: { postId, instagramPostId: igPostId },
    });
  } catch (error) {
    await c.env.DB
      .prepare('UPDATE post_history SET status = ? WHERE id = ?')
      .bind('failed', postId)
      .run();
    throw error;
  }
});

/** GET /api/posts/status/:id - 投稿状態確認 */
post.get('/status/:id', async (c) => {
  const userId = c.get('userId');
  const postId = c.req.param('id');

  const result = await c.env.DB
    .prepare('SELECT id, status, instagram_post_id, posted_at FROM post_history WHERE id = ? AND user_id = ?')
    .bind(postId, userId)
    .first<{
      id: string;
      status: string;
      instagram_post_id: string | null;
      posted_at: string | null;
    }>();

  if (!result) throw new NotFoundError('投稿が見つかりません');
  return c.json({ success: true, data: result });
});

export { post };
